cornac/models/amr/recom_amr.py

```diff
@@ -0,0 +1,298 @@
+        self._init(
+            n_users=train_set.total_users,
+            n_items=train_set.total_items,
+            features=train_features,
+        )
+        
+        if self.trainable:
+            self._fit_torch(train_features)
+        
+        return self
+    
+    def _fit_torch(self, train_features):
+        import torch
+        
+        def _l2_loss(*tensors):
+            l2_loss = 0
+            for tensor in tensors:
+                l2_loss += tensor.pow(2).sum()
+            return l2_loss / 2
+        
+        def _inner(a, b):
+            return (a * b).sum(dim=1)
+        
+        dtype = torch.float
+        device = (
+            torch.device("cuda:0")
+            if (self.use_gpu and torch.cuda.is_available())
+            else torch.device("cpu")
+        )
+        
+        # set requireds_grad=True to get the adversarial gradient
+        # if F is not put into the optimization list of parameters
+        # it won't be updated
+        F = torch.tensor(
+            train_features, device=device, dtype=dtype, requires_grad=True
+        )
+        # Learned parameters
+        Gu = torch.tensor(
+            self.gamma_user, device=device, dtype=dtype, requires_grad=True
+        )
+        Gi = torch.tensor(
+            self.gamma_item, device=device, dtype=dtype, requires_grad=True
+        )
+        E = torch.tensor(
+            self.emb_matrix, device=device, dtype=dtype, requires_grad=True
+        )
+        
+        optimizer = torch.optim.Adam([Gu, Gi, E], lr=self.learning_rate)
+        
+        for epoch in range(1, self.n_epochs + 1):
+            sum_loss = 0.0
+            count = 0
+            progress_bar = tqdm(
+                total=self.train_set.num_batches(self.batch_size),
+                desc="Epoch {}/{}".format(epoch, self.n_epochs),
+                disable=not self.verbose,
+            )
+            for batch_u, batch_i, batch_j in self.train_set.uij_iter(
+                    self.batch_size, shuffle=True
+            ):
+                gamma_u = Gu[batch_u]
+                gamma_i = Gi[batch_i]
+                gamma_j = Gi[batch_j]
+                feat_i = F[batch_i]
+                feat_j = F[batch_j]
+                
+                gamma_diff = gamma_i - gamma_j
+                feat_diff = feat_i - feat_j
+                
+                Xuij = (
+                        _inner(gamma_u, gamma_diff)
+                        + _inner(gamma_u, feat_diff.mm(E))
+                )
+                
+                log_likelihood = torch.nn.functional.logsigmoid(Xuij).sum()
+                
+                # adversarial part
+                feat_i.retain_grad()
+                feat_j.retain_grad()
+                log_likelihood.backward(retain_graph=True)
+                feat_i_delta = feat_i.grad
+                feat_j_delta = feat_j.grad
+                
+                adv_feat_diff = feat_diff - (feat_i_delta - feat_j_delta)
```

Comment:
Hi, why is it `feat_diff - (feat_i_delta - feat_j_delta)`? Based on the equation (5) in the paper, I think it should be `feat_diff + (feat_i_delta - feat_j_delta)`. Thanks!

Reply:
My bad in merging repos.

Should be plus.

@tqtg

Reply:
@RuihongQiu could you create a PR updating the formula?

Reply:
I have made an update on this [PR](https://github.com/PreferredAI/cornac/pull/429).

Please see if it is suitable.

Thank you.

Reply:
Merged! 
The change will be effective in the next release. For up-to-date code, please pull the source directly from GitHub. Thanks!